src/popover-v2/popover-trigger.tsx

```diff
@@ -46,63 +51,46 @@ export const PopoverTrigger = ({
                 limiter: limitShift(),
             }),
         ],
+        onOpenChange: (isOpen) => {
+            setVisible(isOpen);
+
+            // this callback is triggering twice on hover, the check here prevents extra calls
+            if (visible !== isOpen) {
+                handleVisibilityChange(isOpen);
+            }
```

Comment:
why not handle the visibility change in a useEffect? this way you wont need to call it at multiple places / check for previous state

Reply:
oh it will work for this case, just that I generally prefer not to use `useEffect` for events. in a more complex components 1) I might not always want to trigger the effect and 2) it gets hard to trace the originating event

`useEffect` also triggers on mount which requires tracking the first render in this case